drivers/nutdrv_qx.c

```diff
@@ -1911,17 +2036,17 @@ static int	armac_command(const char *cmd, char *buf, size_t buflen)
 	memset(buf, 0, buflen);
 
 	bufpos = 0;
-	while (bufpos + ARMAC_READ_SIZE < buflen) {
+	while (bufpos + read_size + 1 < buflen) {
 		size_t bytes_available;
 
 		/* Read data in 6-byte chunks */
-		ret = usb_interrupt_read(udev, 0x81,
-			(usb_ctrl_charbuf)tmpbuf, ARMAC_READ_SIZE, 1000);
+		ret = usb_interrupt_read(udev, use_interrupt ? armac_endpoint_cache.in_endpoint_address : 0x81,
+			(usb_ctrl_charbuf)tmpbuf, read_size, 1000);
 
 		/* Any errors here mean that we are unable to read a reply
 		 * (which will happen after successfully writing a command
 		 * to the UPS) */
-		if (ret != ARMAC_READ_SIZE) {
+		if (ret != read_size) {
```

Comment:
As we loop here (in current code and the earlier one), would it be less misleading to return not the latest `ret` (if positive, just too short presumably), but a count of successfully read bytes e.g. `bufpos + ret`? If `ret` is negative (for error), probably pass it through as is and disregard any earlier success? Or return `bufpos` anyway, if positive?

Reply:
Nice catch! Fixed via [dc3299e](https://github.com/networkupstools/nut/pull/2154/commits/dc3299e10aec985c8adfd6516dd8580d1f2a5327)

Reply:
@vytautassurvila @blaa : any particular preference for `bufpos` vs. `bufpos + ret` in this case?

Generally I think we can fall into this situation if the device has less info to send than exactly the `read_size`. Like, it has 8 bytes for some message and we read in 6-byte chunks, so our buffer gets `ret=6` in one cycle and `ret=2` in the other quite validly - and if `ret` is positive, we can tell the caller that we have collected 8 bytes.

But this is a general train of thought, you may be better aware about what does and does not happen in these particular interactions :)

Reply:
If I'm not mistaken here we just received data to `tmpbuf`. It will be copied to `buf` only at line 2107. I guess idea is to copy data to `buf` only when some validation is performed on data we received.

Reply:
Right, good catch you too :")

So if we abort here, `buf` has only received `bufpos` finished bytes indeed.

Reply:
I suppose, here and elsewhere we declare that correct reads always return exactly `read_size` and may not be shorter, to the best of our practical knowledge at this time at least?

Reply:
I can speak only about my UPS - `usb_interrupt_read` and `read_size` always match. Within those 64 bytes not all bytes are meaningful - that is determined by bit masking first byte.

Reply:
Ok, thanks for the insights. LGTM I guess. The caller's buffer is large enough for the tested devices, I suppose?